csrc/multidevice/ipc_handle.h

```diff
@@ -0,0 +1,155 @@
+      : local_(std::move(local)), peer_(std::move(peer)) {}
+
+  const auto& local() const {
+    return *local_;
+  }
+
+  const auto& peer() const {
+    return *peer_;
+  }
+
+ private:
+  std::unique_ptr<IpcHandle> local_;
+  std::unique_ptr<IpcHandle> peer_;
+};
+
+// IpcHandleCache manages and cache the IpcHandles.
+// Caching is done on the runtime values of (peer, tensor) and the
+// P2PCommunication* pointer.
+class IpcHandleCache {
+ public:
+  IpcHandleCache(const ExpressionEvaluator& expr_evaluator)
+      : expr_evaluator_(expr_evaluator) {}
+  ~IpcHandleCache() = default;
+
+  // Create IpcHandles, import and export them, and populate the cache. This
+  // method must be called priori to calling get. In many case, the handles need
+  // to be exported by batch (thus the function taking a vector of
+  // P2PCommunication*) to improve performance and to avoid creating deadlocks
+  // when imports and exports order differ accross ranks.
+  void exchangeHandles(const std::vector<P2PCommunication*>& communications);
+
+  // Retrieves a cached item and throws if not present
+  const P2pIpcHandle& get(P2PCommunication* communication) const {
+    auto it = find(communication);
+    NVF_ERROR(
+        it != nullptr,
+        "No remote buffer found for ",
+        communication->toString());
+    return *it;
+  }
+
+ private:
+  using KeyType = std::tuple<int64_t, at::Tensor, P2PCommunication*>;
+
+  KeyType getKey(P2PCommunication* comm) const {
+    int64_t peer = expr_evaluator_.evaluate(comm->peer()).as<int64_t>();
+    at::Tensor buffer =
+        expr_evaluator_.evaluate(comm->buffer()).as<at::Tensor>();
+    return std::make_tuple(peer, buffer, comm);
+  }
+
+  void insert(P2PCommunication* comm, std::unique_ptr<P2pIpcHandle> handle) {
+    handles_[getKey(comm)] = std::move(handle);
+  }
+
+  P2pIpcHandle* find(P2PCommunication* comm) const {
+    auto it = handles_.find(getKey(comm));
+    if (it == handles_.end()) {
+      return nullptr;
+    }
+    return it->second.get();
+  }
+
+  struct TensorHash {
+    std::size_t operator()(const at::Tensor& tensor) const {
+      auto ptr = reinterpret_cast<std::uintptr_t>(tensor.data_ptr());
+      auto offset = tensor.storage_offset();
+      auto element_size = tensor.element_size();
+      return std::hash<std::uintptr_t>()(ptr) ^ std::hash<int64_t>()(offset) ^
+          std::hash<int64_t>()(element_size);
+    }
+  };
+
+  struct TensorEqual {
+    bool operator()(const at::Tensor& lhs, const at::Tensor& rhs) const {
+      return lhs.equal(rhs);
+    }
+  };
+
+  struct KeyHash {
+    std::size_t operator()(const KeyType& key) const {
+      return (std::hash<int64_t>()(std::get<0>(key))) ^
+          (TensorHash{}(std::get<1>(key))) ^
+          (std::hash<P2PCommunication*>()(std::get<2>(key)));
+    }
+  };
+
+  struct KeyEqual {
+    bool operator()(const KeyType& lhs, const KeyType& rhs) const {
+      return std::get<0>(lhs) == std::get<0>(rhs) &&
```

Comment:
Can you make KeyType a struct? This way, readers don't have to remember what 0, 1, and 2 refer to. Also, you can probably define `operator==` there so the equality operator is closer to the definition.